### PR TITLE
Multi user chat

### PR DESCRIPTION
**Pull request by mikkelspring**
Allows admin to create a closed room and register users to get access

### CODE REVIEW

File: src/test/scala/smack/messaging/ClientTest.scala [hunk 1]
```diff
@@ -477,6 +545,109 @@ class ClientTest extends WordSpec with Matchers with BeforeAndAfterEach {
       }
     }
 
+    def createChatRoom = {
+      withChatRoomsActivated {
+        val room = adminUser ? CreateChatRoom(chatRoom)
+        room.value.get shouldBe Success(Created)
+        val rooms = adminUser ? GetChatRooms
+        assert(rooms.value.get.get.asInstanceOf[GetChatRoomsResponse].rooms.contains(fullChatRoom))
```
Comment (mpollmeier) [inline src/test/scala/smack/messaging/ClientTest.scala]: instead of casting yourself, akka has something like 'mappedTo' or 'mapTo'

**Comment by mpollmeier** [comment]
LGTM

**Comment by KadekM** [comment]
I am not fan of this pattern

```
case ex: XMPPErrorException if ex.getXMPPError.getCondition == XMPPError.Condition.forbidden && ex.getXMPPError.getType == ...
case ex: IllegalStateException if ex.getMessage == "Creation failed - User already joined the room." ⇒ ...
case ex: SmackException if ex.getMessage == "Creation failed - Missing acknowledge of room creation." ⇒ ...
...
```
1. exception holding data on which control flow is decided is veeery weird in my opinion. 
2. to make things worse data is actual stringly typed message... 

seems very fragile
Is there not a better way ? I obviously don't have all the context.

**Comment by mikkelspring** [comment]
It’s not really control flow, all it’s doing is translating the internal smack messages to more readable akka failures that are then sent as a message back to sender.

From: Marek Kadek
Reply-To: springnz/smack-scala
Date: Thursday, 15 October 2015 3:12 pm
To: springnz/smack-scala
Cc: Mikkel Fishman
Subject: Re: [smack-scala] Multi user chat (#9)

I am not fan of this pattern

 case ex: IllegalStateException if ex.getMessage == "Creation failed - User already joined the room." ⇒ ...
-              case ex: XMPPErrorException if ex.getXMPPError.getCondition == XMPPError.Condition.forbidden && ex.getXMPPError.getType == ...
-              case ex: IllegalStateException if ex.getMessage == "Creation failed - User already joined the room." ⇒ ...
-              case ex: SmackException if ex.getMessage == "Creation failed - Missing acknowledge of room creation." ⇒ ...
  ...

exception holding data on which control flow is decided is veeery weird in my opinion. Is there not a better way ? I obviously don't have all the context.

**Comment by KadekM** [comment]
That doesn't really make it better. So if client is handling that specific error, and stringly typed message is changed, he will be a bit surprised for getting just the generic message.

It looks a bit weird that now actor turned into 100% receive-respond type of thing -> I don't really see benefit of returning success messages, especially since none of them hold any data.

I personally would delete all ActorSuccess and ActorFailure responses, let exception bubble to parent of Client, and let him handle the situation (whether its resuming operation normally, restarting or killing it all together). Usage of actors like this don't 100% make sense to me! 
_But if it works for you / @mpollmeier thats fine and feel free to ignore me :)_

---

also `uploadAdapter.upload(file, description)` should take ExecutionContext, it should not be hardcoded in implementation of the adapter. Either
a) make adapter not return Future, thus being synchronous, then client knows he needs to start future
b) put EC as argument, so client can specify which ec he wants the future to run on
I prefer (b)

**Pull request merged**